janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphVertexStep.java

```diff
@@ -195,25 +193,53 @@ private void setParentMultiQueryStep() {
         }
 
         if (batchPropertyPrefetching) {
-            Set<Vertex> vertices = new HashSet<>();
-            result.forEach(v -> {
-                if (vertices.size() < txVertexCacheSize ) {
-                    vertices.add((Vertex) v);
-                }
-            });
-
-            // If there are multiple vertices then fetch the properties for all of them in a single multiQuery to
-            // populate the vertex cache so subsequent queries of properties don't have to go to the storage back end
-            if (vertices.size() > 1) {
-                JanusGraphMultiVertexQuery propertyMultiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
-                ((BasicVertexCentricQueryBuilder) propertyMultiQuery).profiler(queryProfiler);
-                propertyMultiQuery.addAllVertices(vertices).preFetch();
-            }
+            prefetchPropertiesIfThereAreMultipleVertices(result);
         }
 
         return (Iterator<E>) result.iterator();
     }
 
+    /**
+     * If there are multiple vertices then fetch the properties for all of them in a single multiQuery to
+     * populate the vertex cache so subsequent queries of properties don't have to go to the storage back end
+     *
+     * @param vertices vertices to check and prefetch. Properties for up to `txVertexCacheSize` first vertices
+     *                 will be prefetched in a single multiQuery. If there is only a single vertex inside `vertices`
+     *                 or `txVertexCacheSize` is less than 2 - no properties will be prefetched for this vertex.
+     */
+    private void prefetchPropertiesIfThereAreMultipleVertices(Iterable<? extends JanusGraphElement> vertices){
+
+        if(txVertexCacheSize < 2){
+            return;
+        }
+
+        Iterator<? extends JanusGraphElement> resultIt = vertices.iterator();
+
+        if(!resultIt.hasNext()){
+            return;
+        }
+
+        JanusGraphElement firstVertex = resultIt.next();
+
+        if(!resultIt.hasNext()){
+            return;
+        }
+
+        JanusGraphMultiVertexQuery propertyMultiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
+        ((BasicVertexCentricQueryBuilder) propertyMultiQuery).profiler(queryProfiler);
+
+        propertyMultiQuery.addVertex((Vertex) firstVertex);
+
+        int addedVertices = 1;
+
+        do {
+            propertyMultiQuery.addVertex((Vertex) resultIt.next());
+            ++addedVertices;
+        } while (addedVertices < txVertexCacheSize && resultIt.hasNext());
```

Comment:
Let's say `txVertexCacheSize == 10`, and there are 15 results. If the first 10 results are identical, `addedVertices` will reach 10 and then break the loop, even though only one unique vertex is included for batch prefetching.

Reply:
Thanks for catching this mistake! You are totally right